### PR TITLE
Add FromVolumeFile time dependent option

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

First in a series of PRs that revamp the time dependent options of the Sphere and BCO domain creators.

Adds an option (not used yet) that can read functions of time that have been serialized in the volume data of an H5 file. The most recent commit of this PR contains the bulk of the changes. The other commits are just some organizational ones and don't contain any real changes.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/Creators/TimeDependentOptions/FromVolumeFile.hpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <brigand/brigand.hpp>
```
Comment (nilsvu) [inline src/Domain/Creators/TimeDependentOptions/FromVolumeFile.hpp]: delete this line

File: src/Domain/Creators/TimeDependentOptions/FromVolumeFile.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+FromVolumeFile<names::Rotation>::FromVolumeFile(
+    const std::string& h5_filename, const std::string& subfile_name,
+    const double time, const bool is_quat_fot,
+    const Options::Context& context) {
+  const std::string function_of_time_name{"Rotation"};
+
+  const auto function_of_time = get_function_of_time(
+      function_of_time_name, h5_filename, subfile_name, time, context);
+
+  if (is_quat_fot) {
```
Comment (nilsvu) [inline src/Domain/Creators/TimeDependentOptions/FromVolumeFile.cpp]: You could get rid of this option by checking the result of the dynamic cast. As a user I wouldn't know what value to choose for this option.

File: src/Domain/Creators/TimeDependentOptions/FromVolumeFile.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    std::array<boost::math::quaternion<double>, 3> quats;
+    for (size_t i = 0; i < quaternions.size(); i++) {
+      gsl::at(quats, i) = datavector_to_quaternion(gsl::at(quaternions, i));
+    }
+
+    // We can't compute the angle so set it to zero
+    angle_values[0] = DataVector{4, 0.0};
+    angle_values[1] = quaternion_to_datavector(2.0 * conj(quats[0]) * quats[1]);
+    angle_values[2] = quaternion_to_datavector(
+        conj(quats[0]) * (2.0 * quats[2] - quats[1] * datavector_to_quaternion(
```
Comment (nilsvu) [inline src/Domain/Creators/TimeDependentOptions/FromVolumeFile.cpp:169]: where do these computations come from? Please add a comment point to some docs

File: tests/Unit/Domain/Creators/TimeDependentOptions/Test_FromVolumeFile.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  }
+
+  {
+    INFO("IsQuaternionFunctionOfTime = False");
+    // Going at t=0 is easier for checking quaternions
+    const auto from_volume_file = TestHelpers::test_creation<
+        domain::creators::time_dependent_options::FromVolumeFile<
+            domain::creators::time_dependent_options::names::Rotation>>(
+        "H5Filename: " + filename + "\nSubfileName: " + subfile_name +
+        "\nTime: 0.0\nIsQuaternionFunctionOfTime: False\n");
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/TimeDependentOptions/Test_FromVolumeFile.cpp]: Doesn't this read the same quat FoT as above?

File: tests/Unit/Domain/Creators/TimeDependentOptions/Test_FromVolumeFile.cpp [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    file_system::rm(filename, true);
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.Domain.Creators.TimeDependentOptions.FromVolumeFile",
+                  "[Unit][Domain]") {
+  domain::FunctionsOfTime::register_derived_with_charm();
+  test<domain::creators::time_dependent_options::names::Translation>();
+  test<domain::creators::time_dependent_options::names::Expansion>();
+  test<domain::creators::time_dependent_options::names::Rotation>();
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/TimeDependentOptions/Test_FromVolumeFile.cpp:357]: This is missing shape/size, no?

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by knelli2** [comment]
@nilsvu Because of discussions over the past few days about reading in the FoTs exactly, I added a new commit that exposes all derivatives of a FoT. I then use that for the rotation and size FoTs in a fixup commit.

**Comment by nilsvu** [comment]
squash

**Review by nilsvu — APPROVED** [review]

**Pull request merged**